Ole32 project

vatsanm opened the pull request:
- Added a project for Ole32, along with PInvokes for core COM call cancellation API's
- No tests added. 

Question: Should CoEnableCallCancellation(IntPtr) and CoDisableCallCancellation(IntPtr) be made private, in favor of exclusively exposing the helper metohds CoEnableCallCancellation(void) and CoDisableCallCancellation(void) respectively ?

vatsan-madhavan committed:
Added a project for Ole32, along with PInvokes for core COM call cancelation API's, and related HResult codes.

vbfox commented on src/Ole32/Ole32.csproj:
```diff
@@ -0,0 +1,9 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>$(DesktopAndPortableFrameworks)</TargetFrameworks>
```
As far as I can see It should be `DesktopOnlyFrameworks` old32 doesn't seem supported on non-desktop no ?

vatsanm replied:
IIRC Ole32 is supported on Windows Phone, but I'm not quite sure if it's supported on, say, Windows Server Nano. Even on phone, I don't think that all Ole32 functions are available. I'm not quite sure what the delta between DesktopAndPortableFrameworks and DesktopOnlyFrameworks  is though. Are there other questions I could answer (or research the answers for) that would help us be sure that we are making the right choice?

vbfox replied:
I think @AArnott will have the definitive answer on that.

My rule of thumb was that that Portable = UWP and so that it would never run on desktop.
(It also seem to be what our other project do)

But it's true that a netstandard dll could be run by dotnet core on windows and Ole32 WILL be there, but I don't think we support this case anywhere.

Ideally we would build our current "desktop" assemblies for everything including netstandard. And special-case the `uap` platform with what is currently our "portable" set.

But it's not what the current props file seem to do https://github.com/AArnott/pinvoke/blob/master/src/Directory.Build.props

AArnott replied:
At least one method you added is MSDN documented as `Windows 2000 Professional [desktop apps only]`. So we should either target just Desktop profiles here, or carefully move the right subset of methods to a desktop-only source file as we've done in other projects (see Kernel32 as an example).

> But it's true that a netstandard dll could be run by dotnet core on windows and Ole32 WILL be there, but I don't think we support this case anywhere.

True. And if pressed upon by anyone, we could add the desktop-only APIs to the portable SKUs for the sake of dotnet.exe running on Windows. But we'd have to (as @vbfox said) compile `uap` to exclude it otherwise users would get store certification failures since it doesn't allow defining these desktop-only methods (even if unused). 
But for now, since no one is asking for it, the simplest and current policy is to make them only available when targeting desktop specifically. 

So yes, please change this to `$(DesktopOnlyFrameworks)`.

vatsanm replied:
Alright, done.

AArnott commented on src/Ole32/Ole32.Helpers.cs:
```diff
@@ -0,0 +1,53 @@
+        /// <returns>This function can return the standard return values <see cref="HResult.Code.S_OK"/>, <see cref="HResult.Code.E_FAIL"/>, <see cref="HResult.Code.E_INVALIDARG"/>, and <see cref="HResult.Code.E_OUTOFMEMORY"/>.</returns>
+        public static HResult CoEnableCallCancellation()
+        {
+            return CoEnableCallCancellation(IntPtr.Zero);
```
We don't typically offer helper methods that simply remove a parameter. We generally reserve them for assisting with difficult memory management. 
What if we remove these two helper methods and add `= default(IntPtr)` to the `[dllimport]` methods themselves? It should accomplish the same thing.

AArnott replied:
Thanks. Did you want to use the `= null` default parameter approach?

AArnott commented on src/Ole32/Ole32.cs:
```diff
@@ -0,0 +1,82 @@
+        /// <param name="pReserved">This parameter is reserved and must be NULL.(<see cref="IntPtr.Zero"/>)</param>
+        /// <returns>This function can return the standard return values <see cref="HResult.Code.S_OK"/>, <see cref="HResult.Code.E_FAIL"/>, <see cref="HResult.Code.E_INVALIDARG"/>, and <see cref="HResult.Code.E_OUTOFMEMORY"/>.</returns>
+        [DllImport(nameof(Ole32))]
+        public static extern HResult CoEnableCallCancellation(IntPtr pReserved);
```
I think strictly speaking this should be `void*` (and we'll automatically generate `IntPtr` overloads).

AArnott commented on src/Ole32/Ole32.cs:
```diff
@@ -0,0 +1,82 @@
+        ///     </item>
+        /// </list></returns>
+        [DllImport(nameof(Ole32))]
+        public static extern HResult CoCancelCall(uint dwThreadId, uint ulTimeout);
```
We generally prefer to use `int` instead of `uint`, especially when it's likely that the user will want to pass in a variable that came from .NET. For example, .NET uses `int` for thread IDs, and if we use `uint` here it will require a cast. I would tend to say the same about your `ulTimeout` parameter. I suggest we make them both `int`.

AArnott reviewed (changes requested):
Thanks so much for the contribution. It looks like you were fairly meticulous to follow our published guidance, and we appreciate it.

vatsan-madhavan committed:
* Changing Ole32.csproj to target $(DesktopOnlyFrameworks) instead of $(DesktopAndPortableFrameworks)
* Removing Ole32 helper methods
* Changing Ole32 methods currently taking IntPtr parameters to instead take void* parameters. This lets codegen to take care of auto-generating the corresponding IntPtr version.
* Fixing up Ole32.Unshipped.txt appropriately.

vatsanm commented:
Addressed feedback. 

The one thing I couldn't do was add `=default(IntPtr)` or `void* = null`. When I tried doing this, I got the following CodeAnalysis warning: 

> Ole32.cs(22,45): warning RS0027: Symbol 'CoEnableCallCancellation' violates the backcompat requirement: 'Public API with optional parameter(s) should have the most parameters amongst its public overloads'. See 'https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md' for details. [D:\src\repos\pinvoke\src\Ole32\Ole32.csproj]

I suspect this warning occurs because the code-generator isn't currently transfering over default arguments (i..e, it is not converting `void* pReserved = null` to `System.IntPtr pReserved = IntPtr.Zero`), and the resultant generated code simply skips the addition of a default value. 

Optional argument/default value would have been nice to have given that the only legal value for these arguments in `CoEnableCallCancellation` and `CoDisableCallCancellatoin` is `null`/`IntPtr.Zero`. That said, this is really only a minor drawback.
